tests/e2e-playwright/utils/wait_for_services.py

```diff
@@ -0,0 +1,180 @@
+    datetime_str = service.attrs["CreatedAt"]
+    return to_datetime(datetime_str)
+
+
+def wait_for_services() -> int:
+    expected_services = core_services() + ops_services()
+    started_services = []
+    client = docker.from_env()
+    try:
+        for attempt in Retrying(
+            stop=stop_after_delay(MAX_WAIT_TIME),
+            wait=wait_fixed(WAIT_BEFORE_RETRY),
+            before_sleep=before_sleep_log(logger, logging.WARNING),
+        ):
+            with attempt:
+                started_services = sorted(
+                    (
+                        s
+                        for s in client.services.list()
+                        if s.name.split("_")[-1] in expected_services
+                    ),
+                    key=by_service_creation,
+                )
+
+                assert len(started_services), "no services started!"
+                assert len(expected_services) == len(started_services), (
+                    "Some services are missing or unexpected:\n"
+                    f"expected: {len(expected_services)} {expected_services}\n"
+                    f"got: {len(started_services)} {[s.name for s in started_services]}"
+                )
+    except RetryError:
+        print(
+            f"found these services: {len(started_services)} {[s.name for s in started_services]}\nexpected services: {len(expected_services)} {expected_services}"
+        )
+        return os.EX_SOFTWARE
+
+    for service in started_services:
+        expected_replicas = (
+            service.attrs["Spec"]["Mode"]["Replicated"]["Replicas"]
+            if "Replicated" in service.attrs["Spec"]["Mode"]
+            else len(client.nodes.list())  # we are in global mode
+        )
+        print(f"Service: {service.name} expects {expected_replicas} replicas", "-" * 10)
+
+        try:
+            for attempt in Retrying(
+                stop=stop_after_delay(MAX_WAIT_TIME),
+                wait=wait_fixed(WAIT_BEFORE_RETRY),
+            ):
+                with attempt:
+                    service_tasks: list[dict] = service.tasks()  #  freeze
+                    print(get_tasks_summary(service_tasks))
+
+                    #
+                    # NOTE: a service could set 'ready' as desired-state instead of 'running' if
+                    # it constantly breaks and the swarm desides to "stop trying".
+                    #
+                    valid_replicas = sum(
+                        task["Status"]["State"] == RUNNING_STATE
+                        for task in service_tasks
+                    )
+                    assert valid_replicas == expected_replicas
+        except RetryError:
+            print(
+                f"ERROR: Service {service.name} failed to start {expected_replicas} replica/s"
+            )
+            print(json.dumps(service.attrs, indent=1))
+            return os.EX_SOFTWARE
+
+    return os.EX_OK
+
+
+if __name__ == "__main__":
+    sys.exit(wait_for_services())
```

Comment:
is this used to test manually?

Reply:
I see that this this script is used directly by the Makefile

Reply:
this script is used by the Makefile yes.